gtdynamics/optimizer/ConstraintManifold.cpp

```diff
@@ -0,0 +1,275 @@
+#include <gtsam/linear/VectorValues.h>
+#include <gtsam/nonlinear/LinearContainerFactor.h>
+
+namespace gtsam {
+
+/* ************************************************************************* */
+void ConstraintManifold::compute_values(const gtsam::Values& values,
+                                        const bool retract_init) {
+  values_ = gtsam::Values();
+  base_dim_ = 0;
+  for (const gtsam::Key& key : cc_->keys) {
+    const auto& value = values.at(key);
+    base_dim_ += value.dim();
+    values_.insert(key, value);
+  }
+  constraint_dim_ = 0;
+  for (const auto& constraint : cc_->constraints) {
+    constraint_dim_ += constraint->dim();
+  }
+  if (base_dim_ > constraint_dim_) {
+    dim_ = base_dim_ - constraint_dim_;
+  } else {
+    dim_ = 0;
+  }
+  if (retract_init) {
+    values_ = retract_constraints(values_);
+  }
+}
+
+/* ************************************************************************* */
+Values ConstraintManifold::retract_constraints(const Values& values) const {
+  if (params_->retract_type == Params::RetractType::UOPT) {
+    return retract_uopt(values);
+  } else if (params_->retract_type == Params::RetractType::PROJ) {
+    return retract_proj(values);
+  } else if (params_->retract_type == Params::RetractType::PARTIAL_PROJ) {
+    return retract_p_proj(values);
+  }
+  std::cerr << "Unrecognized retraction type.\n";
+  return Values();
+}
+
+/* ************************************************************************* */
+void ConstraintManifold::compute_basis() {
+  if (params_->basis_type == Params::BasisType::KERNEL) {
+    compute_basis_kernel();
+  } else if (params_->basis_type == Params::BasisType::SPECIFY_VARIABLES) {
+    compute_basis_specify_variables();
+  }
+}
+
+/* ************************************************************************* */
+const gtsam::Value& ConstraintManifold::recover(const gtsam::Key key,
+                                                ChartJacobian H) const {
+  if (H) {
+    // choose the corresponding rows in basis
+    *H = basis_.middleRows(var_location_.at(key), var_dim_.at(key));
+  }
+  return values_.at(key);
+}
+
+/* ************************************************************************* */
+ConstraintManifold ConstraintManifold::retract(const gtsam::Vector& xi,
+                                               ChartJacobian H1,
+                                               ChartJacobian H2) const {
+  // Compute delta for each variable and perform update.
+  gtsam::Vector x_xi = basis_ * xi;
+  gtsam::VectorValues delta;
+  for (const Key& key : cc_->keys) {
+    delta.insert(key, x_xi.middleRows(var_location_.at(key), var_dim_.at(key)));
+  }
+  gtsam::Values new_values = values_.retract(delta);
+
+  // Temporarily set jacobian as 0 since they are not used for optimization.
+  if (H1) H1->setZero();
+  if (H2) H2->setZero();
+
+  // Satisfy the constraints in the connected component.
+  return createWithNewValues(new_values, true);
```

Comment:
> just checking my understanding, am I correct in saying this would be the same as:
> 
> ```c++
> new_values = retract_constraints(new_values);
> return createWithNewValues(new_values, false);
> ```
> 
> or is it different because in the version as written the new initialization point is used instead of the old?

Correct!

Reply:
Done. Note that `retract` is function required for all manifold (Value) type in gtsam, while retract_cosntraints provides the detailed implementation of each retract option.